src/fields/field2_128/extend.rs

```diff
@@ -0,0 +1,355 @@
+    // = TWIDDLE(i, coset + (u + 2^k) * 2^(curr_power + 1)) (by †)
+    // = TWIDDLE(i, coset + u * 2^(curr_power + 1) + 2^k * 2^(curr_power + 1)
+    // = TWIDDLE(i, coset + u * 2^(curr_power + 1) + 2^(k + curr_power + 1))
+    // = TWIDDLE(i, coset + u * 2^(curr_power + 1)) + W_hat[curr_power][k + curr_power + 1] (by *)
+    // = TWIDDLE[u] + W_hat[curr_power][k + curr_power + 1] (by †)
+    twiddled[0] = twiddle(curr_power, coset);
+    for k in 0..(power - curr_power - 1) {
+        for u in 0..1 << k {
+            twiddled[u + (1 << k)] = twiddled[u] + twiddle_array_at(curr_power, k + curr_power + 1);
+        }
+    }
+}
+
+/// Implements procedure BUTTERFLY-FWD from Algorithm 1 in section 3.2 of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft_butterfly_forward(
+    fft_array: &mut [Field2_128],
+    index: usize,
+    recursive_len: usize,
+    twiddle: Field2_128,
+) {
+    fft_array[index] += twiddle * fft_array[index + recursive_len];
+    fft_array[index + recursive_len] += fft_array[index];
+}
+
+/// Implements procedure BUTTERFLY-BWD from Algorithm 1 in section 3.2 of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft_butterfly_backward(
+    fft_array: &mut [Field2_128],
+    index: usize,
+    recursive_len: usize,
+    twiddle: Field2_128,
+) {
+    fft_array[index + recursive_len] -= fft_array[index];
+    fft_array[index] -= twiddle * fft_array[index + recursive_len];
+}
+
+/// Implements procedure BUTTERFLY-DIAG from Algorithm 1 in section 3.2 of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft_butterfly_diagonal(
+    fft_array: &mut [Field2_128],
+    index: usize,
+    recursive_len: usize,
+    twiddle: Field2_128,
+) {
+    let prev_at_index = fft_array[index];
+
+    fft_array[index] -= twiddle * fft_array[index + recursive_len];
+    fft_array[index + recursive_len] += prev_at_index;
+}
+
+/// Direction in which the FFT operates.
+#[derive(Eq, PartialEq)]
+enum Direction {
+    Forward,
+    Backward,
+}
+
+/// Implements procedure FFT and IFFT (depending on direction) from Algorithm 1 in section 3.2
+/// of [the paper][1].
+///
+/// [1]: https://eprint.iacr.org/2024/2010.pdf
+fn fft(direction: Direction, power: u32, coset: usize, fft_array: &mut [Field2_128]) {
+    if power == 0 {
+        return;
+    }
+    let mut twiddled = vec![Field2_128::ZERO; 1 << (power - 1)];
+    for mut curr_power in 0..power {
+        // Forward FFT iterates over power..0
+        if direction == Direction::Forward {
+            curr_power = power - curr_power - 1;
+        }
+        let recursive_len = 1 << curr_power;
+
+        twiddles(curr_power, power, coset, &mut twiddled);
+
+        // for all u : 0 ≤ 2s · u < 2ℓ
+        for (index, start) in (0..1 << power).step_by(2 * recursive_len).enumerate() {
+            let twiddle = twiddled[index];
+            for v in 0..recursive_len {
+                match direction {
```

Comment:
I'm curious if manual loop unswitching could speed this up, but that will require benchmarking to answer.